### PR TITLE
add feature importance for C45

### PR DESCRIPTION
**Pull request by LiuYuHui**
if it look fine, i will update meta example and data :)

### CODE REVIEW

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (vigsterkr) [inline src/shogun/multiclass/tree/C45ClassifierTree.cpp]:
```suggestion
	#pragma omp parallel for shared(m_feature_importances)
	for (const auto& child : childs)
```

not so sure if its worth it though...

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (vigsterkr) [inline src/shogun/multiclass/tree/C45ClassifierTree.cpp]:
```suggestion
	SG_ADD(
	    &m_feature_importances, "feature_importances", "feature importances");
```

Comment (vigsterkr) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: just to be consistent with the other trees

**Review by vigsterkr — COMMENTED** [review]
some minor things

**Comment by vigsterkr** [comment]
yep lets have the meta examples + data added if possible!

have you actually tested that omp on a bigger tree? :D it should work, but 🤷‍♂

**Comment by LiuYuHui** [comment]
> 
> 
> yep lets have the meta examples + data added if possible!
> 
> have you actually tested that omp on a bigger tree? :D it should work, but 🤷‍♂

maybe i should use benchmark to compare these two situations :D

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (gf712) [inline src/shogun/multiclass/tree/C45ClassifierTree.cpp]: This should either be registered with watch_method or pass the READONLY flag

**Review by gf712 — CHANGES_REQUESTED** [review]

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (LiuYuHui) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: this ? `SG_ADD(
	    &m_feature_importances, "feature_importances", "feature importances",ParameterProperties::READONLY);`

Comment (gf712) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: yes!

Comment (vigsterkr) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: i've suggested @LiuYuHui that actually maybe it'd be better to unify this for trees... meaning the feature importance...

Comment (LiuYuHui) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: this pr or other pr?

Comment (vigsterkr) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: just do it (here)! (tm)

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (karlnapf) [inline src/shogun/multiclass/tree/C45ClassifierTree.cpp]: minor: grammar ("it" is singular but importances are plural)

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (karlnapf) [inline src/shogun/multiclass/tree/C45ClassifierTree.cpp]: Can it be zero?

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (karlnapf) [inline src/shogun/multiclass/tree/C45ClassifierTree.cpp]: minor: childs -> children

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (karlnapf) [inline src/shogun/multiclass/tree/C45ClassifierTree.cpp]: is this a thread-safe (const method)?

Comment (karlnapf) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: just checked it is not, then the question is: is the multithreading safe?

File: src/shogun/multiclass/tree/C45ClassifierTree.h
Comment (karlnapf) [inline src/shogun/multiclass/tree/C45ClassifierTree.h]: minor: remove first line of comment, redundant

File: src/shogun/multiclass/tree/C45ClassifierTree.cpp
Comment (vigsterkr) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: hence shared(m_feature_importances)
but @LiuYuHui is actually working on some ~unification~ generalization atm

Comment (karlnapf) [reply src/shogun/multiclass/tree/C45ClassifierTree.cpp]: I guess shared only protects writing to the array?
but ok, lets see the unified version

File: src/shogun/multiclass/tree/FeatureImportanceTree.h
Comment (vigsterkr) [inline src/shogun/multiclass/tree/FeatureImportanceTree.h]: Maybe `FeatureImportanceMixin` would be a better name?
@gf712 @karlnapf ?

Comment (gf712) [reply src/shogun/multiclass/tree/FeatureImportanceTree.h]: Yes, and change the typename to something meaningful, e.g TreeType

Comment (karlnapf) [reply src/shogun/multiclass/tree/FeatureImportanceTree.h]: I think having “mixin” in the name is not necessary as it is bulky, would rather put that in the api docstring. But +1 for a meaningful type name

**Comment by gf712** [comment]
Thank you, this LGTM. Maybe let's wait until Azure Pipelines is back and then merge?

File: src/shogun/multiclass/tree/FeatureImportanceTree.h
Comment (gf712) [inline src/shogun/multiclass/tree/FeatureImportanceTree.h]: didn't you use omp here in a previous commit?

Comment (LiuYuHui) [reply src/shogun/multiclass/tree/FeatureImportanceTree.h]: omp seems like don't support with range for

Comment (gf712) [reply src/shogun/multiclass/tree/FeatureImportanceTree.h]: Ah, that's easy to fix, you just need have a oldschool for loop `for(int i =0; i < children.size(); ++i){auto& child = children[i]; ...}`. If the computation is quite expensive you could just use std::thread instead tbh. What do you think @vigsterkr?

File: src/shogun/multiclass/tree/FeatureImportanceTree.h
Comment (gf712) [inline src/shogun/multiclass/tree/FeatureImportanceTree.h]: Const?

Comment (gf712) [reply src/shogun/multiclass/tree/FeatureImportanceTree.h]: Did you also see any performance improvement with omp?

Comment (LiuYuHui) [reply src/shogun/multiclass/tree/FeatureImportanceTree.h]: i have tested, but it seems like that there is no significant performance improvement

**Comment by vigsterkr** [comment]
@LiuYuHui i guess the only thing is missing adding the shogun-data, right? could you update the data submodule (i've just merged your PR to shogun-data) and then we can get this merged! great stuff!

**Comment by karlnapf** [comment]
:rocket:

**Comment by karlnapf** [comment]
quick question. How do we know this implementation is correct and no error slipped in? I am always paranoid about that possibility (happened before) Maybe we could have a unit test? Or you think that is not necessary?

**Comment by vigsterkr** [comment]
@karlnapf i was under the impression (i was told) that @LiuYuHui worked based on the scikitlearn implementation... hoping that he has actually checked its correctness :)

**Comment by karlnapf** [comment]
Maybe we can open an entrance task that writes a unit test that compares numbers against sklearn? Or @LiuYuHui could do it? It also covers future refactors.  Just in case something slipped or might slip in the future.

**Comment by gf712** [comment]
The unit tests use sklearn as a reference!

**Comment by karlnapf** [comment]
but do we have one that checks the importances that were added?

**Comment by vigsterkr** [comment]
you mean the toy data based?

**Comment by LiuYuHui** [comment]
> 
> 
> Maybe we can open an entrance task that writes a unit test that compares numbers against sklearn? Or @LiuYuHui could do it? It also covers future refactors. Just in case something slipped or might slip in the future.

i have added CARTree unit test for feature importance, but it seems that sklearn just implement CARTree, and there is no C45 and id3 implementation in sklearn, so i haven't added C45 and id3 unittest

**Comment by karlnapf** [comment]
A unit test of some corner cases where you can compute the results manually in a simple case would also do it. Just anything really that serves as a minimal line of defense against unintended changes later on. Or, alternatively another framework?

**Pull request merged**
